### PR TITLE
[SPARK-6494][Core]rdd polymorphic  method zipPartitions  refactor

### PR DESCRIPTION
**Pull request by shijinkui**
no need so many polymorphic method, only add default value instead.

modify partition.size instead of partition.length, partitions is Array object.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 1]
```diff
@@ -17,36 +17,28 @@
...
 import org.apache.spark.deploy.SparkHadoopUtil
 import org.apache.spark.executor.DataReadMethod
 import org.apache.spark.rdd.HadoopRDD.HadoopMapPartitionsWithSplitRDD
-import org.apache.spark.util.{NextIterator, Utils}
-import org.apache.spark.scheduler.{HostTaskLocation, HDFSCacheTaskLocation}
+import org.apache.spark.scheduler.{HDFSCacheTaskLocation, HostTaskLocation}
 import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.{NextIterator, Utils}
+
+import scala.collection.immutable.Map
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: These imports are now out of order.

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 2]
```diff
@@ -99,27 +93,27 @@ private[spark] class HadoopPartition(rddId: Int, idx: Int, @transient s: InputSp
  */
 @DeveloperApi
 class HadoopRDD[K, V](
+  sc: SparkContext,
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: This indentation is incorrect now. I would not change imports/indentation unless you know it's wrong

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 3]
```diff
@@ -185,8 +181,10 @@ class HadoopRDD[K, V](
     // done in each local process.
     val newInputFormat = ReflectionUtils.newInstance(inputFormatClass.asInstanceOf[Class[_]], conf)
       .asInstanceOf[InputFormat[K, V]]
-    if (newInputFormat.isInstanceOf[Configurable]) {
-      newInputFormat.asInstanceOf[Configurable].setConf(conf)
+    newInputFormat match {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: These are cases where I think `match` statements can make things a bit less clear

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 4]
```diff
@@ -231,11 +231,11 @@ class HadoopRDD[K, V](
       var reader: RecordReader[K, V] = null
       val inputFormat = getInputFormat(jobConf)
       HadoopRDD.addLocalConfiguration(new SimpleDateFormat("yyyyMMddHHmm").format(createTime),
-        context.stageId, theSplit.index, context.attemptNumber, jobConf)
+        context.stageId(), theSplit.index, context.attemptNumber(), jobConf)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: There are a number of trivial changes like adding parents to what are probably logically no-side-effect methods, and extra whitespaces

File: core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala [hunk 5]
```diff
@@ -60,7 +60,7 @@ private[spark] abstract class ZippedPartitionsBaseRDD[V: ClassTag](
       val prefs = rdds.map(rdd => rdd.preferredLocations(rdd.partitions(i)))
       // Check whether there are any hosts that match all RDDs; otherwise return the union
       val exactMatchLocations = prefs.reduce((x, y) => x.intersect(y))
-      val locs = if (!exactMatchLocations.isEmpty) exactMatchLocations else prefs.flatten.distinct
+      val locs = if (exactMatchLocations.nonEmpty) exactMatchLocations else prefs.flatten.distinct
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala]: This change actually breaks the logic

**Comment by srowen** [comment]
Most of this pull request is wrong, including introducing a logic error. The bits that might be of marginal value are probably too trivial to bother with, like using `length` instead of `size` or cleaning up with a `match` statement. That is, I'm not sure what's left is worth doing. Do you mind closing this?

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 3 above)
Comment (shijinkui) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: maybe.
i think too much nest if statement not looking better.

**Comment by shijinkui** [comment]
@srowen thanks for viewing the trivial changes. the `zipPartitions` maybe have litter value.

when readding code, i fount these trivial place, modify it by the way. i'm not bored:)

just like you said marginal value, but when reading code in IDE, you can see so many yellow tips below screen shot. bad smell. if not clean the bad smell code, it will be everywhere.  

if not aggree, i'll close it. 
Thanks 

screen shot 2015-03-25 at 7 50 48 am

File: core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala (hunk 5 above)
Comment (shijinkui) [inline core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala]: `def isEmpty: Boolean = lengthCompare(0) == 0`     
`def nonEmpty: Boolean = !isEmpty`

`!exactMatchLocations.isEmpty`  equal  `exactMatchLocations.nonEmpty`

nonEmpty maybe more clear semantically

**Comment by shijinkui** [comment]
@srowen @rxin  please review zipPartitions refactor only.
thanks

**Comment by rxin** [comment]
Thanks for looking at this, @shijinkui. Unfortunately, this actually breaks binary compatibility of the API, and thus cannot be merged.

**Comment by shijinkui** [comment]
@rxin , curry function's first parameter will use default value if you miss one value. it will be compatible  ：）

test code like this:
```
object Test11 {

  def main(args: Array[String]): Unit = {
    a1(1)(3)
    a1(1, 2)(3)
  }

  def a1(a: Int, b: Int = 0)(c: Int) = {
    println(a + "," + b + "," + c)
  }
}
```

result:
1,0,3
1,2,3

**Comment by sryza** [comment]
@shijinkui your example demonstrates that *source* compatibility can be preserved, but can *binary* compatibility?

File: core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala (hunk 5 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/ZippedPartitionsRDD.scala]: Oh I misread the original version. Yes that's equivalent. I don't know if it's worth changing, but `nonEmpty` is cleaner for new code.

**Comment by shijinkui** [comment]
@sryza @rxin  understand, thx

**Pull request closed without merging**